# Calculate Basketball Scoring Probabilities
___

## About Dataset

This dataset opens the door to the intricacies of the 2023 NBA season, offering a profound understanding of the art of scoring in professional basketball. Within its comprehensive analysis, it showcases the remarkable prowess of 3 players LeBron James, James Harden, and Stephen Curry—true icons of the sport. Delve deep into the strategic brilliance that defines these players' shooting trends, performance metrics, and unwavering precision on the court. Whether you're a passionate basketball enthusiast or a data-driven analyst, this dataset provides a unique and invaluable window into the mastery of these legendary athletes and the ever-evolving game of basketball.


### Steph Curry Shot Dataset

| Dataset Columns | Description |
| --- | --- |
| **top** | The vertical position on the court where the shot was taken. |
| **left** | The horizontal position on the court where the shot was taken. |
| **date** | The date when the shot was taken (e.g., Oct 18, 2022). |
| **qtr** | The quarter of the attempt (e.g., "1st Qtr", "2nd Qtr"). |
| **time_remaining** | Time remaining in the quarter (e.g., 09:26). |
| **result** | Success of the shot: **TRUE** (made) or **FALSE** (missed). |
| **shot_type** | Type of shot: **2** (two-pointer) or **3** (three-pointer). |
| **distance_ft** | Distance in feet from the hoop to the shot location. |
| **lead** | **TRUE** if Curry's team was leading, **FALSE** otherwise. |
| **player_team_score** | Steph Curry's team's score at the time of the shot. |
| **opponent_team_score** | The opposing team's score at the time of the shot. |
| **opponent** | Abbreviation for the opposing team (e.g., LAL). |
| **team** | Abbreviation for Steph Curry's team (e.g., GSW). |
| **season** | The season year (e.g., 2023). |
| **color** | Color code (e.g., "red" or "green") for visual shot mapping. |

---

### How to use this dataset

Data Scientists and Analysts: Employ advanced statistical analysis to uncover hidden patterns and insights in the shooting trends of LeBron James, James Harden, and Stephen Curry.

## Purpose of this report is analyzing the statistics for Stepy Curry

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings

warnings.filterwarnings("ignore", category=FutureWarning)

# --- Configurações gerais ---
pd.set_option("display.max_columns", None)
pd.set_option("display.float_format", lambda x: f"{x:.2f}")

# --- Estilo dos gráficos ---
sns.set_theme(style="whitegrid", palette="muted")
plt.rcParams["figure.figsize"] = (12, 6)
plt.rcParams["figure.dpi"] = 120
plt.rcParams["axes.titlesize"] = 14
plt.rcParams["axes.labelsize"] = 12

In [2]:
df = pd.read_csv("/home/fabiolima/Workdir/interviews/andela_prep/statistics_review/data/stephen_curry_shots_2023.csv")

In [3]:
print(f"Dataset loaded with {df.shape[0]} rows and {df.shape[1]} columns.")

Dataset loaded with 1434 rows and 15 columns.


## Dataset Overview

- **Total shots:** 1,434
- **No missing values** — data is clean and ready for analysis

### Key columns for this project

| Column | Type | Description |
|---|---|---|
| `result` | bool | True = make, False = miss |
| `shot_type` | int | 2 or 3 pointer |
| `lead` | bool | True = team was leading when shot was taken |

---

## Concept: Empirical Probability

Instead of a theoretical model, we estimate probability directly from observed data:

$$P(A) = \frac{\text{count of event A}}{\text{total observations}}$$

---

## Your Task

Calculate and store these four values as variables — they will be the **building blocks** for every calculation in the rest of the project:

1. $P(\text{make})$ — proportion of `result == True`
2. $P(\text{miss})$ — complement: $1 - P(\text{make})$
3. $P(\text{3-pointer})$ — proportion of `shot\_type == 3`
4. $P(\text{2-pointer})$ — proportion of `shot\_type == 2`

---

## Python Hint

```python
total = len(df)

p_make = ...
p_miss = ...
p_three = ...
p_two = ...
```

> Store these as variables — you will reuse them constantly in Sections 2, 3, and 4.

In [4]:
# Calculating the total number of shots
total_shots = df.shape[0]
print(f"Total number of shots: {total_shots}")

Total number of shots: 1434


## Section 1 — Formula

**Empirical (frequentist) probability** — estimated directly from data:

$$P(A) = \frac{\text{count}(A)}{N}$$

where $N$ is the total number of observations.

Applied here:

$$P(\text{make}) = \frac{\text{count}(\text{result}=\text{True})}{N}$$

$$P(\text{miss}) = 1 - P(\text{make})$$

$$P(\text{3-pointer}) = \frac{\text{count}(\text{shot\_type}=3)}{N}$$

$$P(\text{2-pointer}) = \frac{\text{count}(\text{shot\_type}=2)}{N}$$

## Section 1 - Overall Shooting Statistics
___

* Overall probability of a make (result = TRUE)
* Overall probability of a miss (result = FALSE)
* Overall proportion of three-pointers (shot_type = 3)
* Overall proportion of two-pointers (shot_type = 2)

In [5]:
# Overall probability of make (result = True)
p_make_percentage = df["result"].mean()
p_make = df["result"].sum()/total_shots
print(f"Overall probability of making a shot: {p_make:}")


Overall probability of making a shot: 0.4902370990237099


In [6]:
# Overall probability of miss (result = False)
p_miss_percentage = 1 - p_make_percentage
p_miss = 1 - p_make
print(f"Overall probability of missing a shot: {p_miss}")

Overall probability of missing a shot: 0.5097629009762901


* **Overall proportion of two-points(shot_type=2)**
* **Overall proportion of three-points(shot_type=3)**

In [7]:
# Count of 2s and 3 type-shots: count_two and count_three
count_shot_type = df['result'].groupby(df['shot_type']).count()
print(f"Count of 2-point shots: {count_shot_type[2]}")
print(f"Count of 3-point shots: {count_shot_type[3]}")

Count of 2-point shots: 648
Count of 3-point shots: 786


**The probabilities can be calculate as follow**:

$$P = \frac{\text{event count}}{\text{total}}$$

In [8]:
p_two = count_shot_type[2] / total_shots
p_three = count_shot_type[3] / total_shots
print(f"Probability of shot-type 2: {p_two:.4f}")
print(f"Probability of shot-type 3: {p_three:.4f}")

Probability of shot-type 2: 0.4519
Probability of shot-type 3: 0.5481


In [9]:
print("=" * 45)
print("Section 1: OVERALL SHOOTING STATISTICS")
print("=" * 45)
print(f"{'Metric':<30} {'Probability':>10}")
print("-" * 45)
print(f"{'P(Make)':<30} {p_make:>10.4f}")
print(f"{'P(Miss)':<30} {p_miss:>10.4f}")
print(f"{'P(Three-pointer)':<30} {p_three:>10.4f}")
print(f"{'P(Two-pointer)':<30} {p_two:>10.4f}")
print("=" * 45)

Section 1: OVERALL SHOOTING STATISTICS
Metric                         Probability
---------------------------------------------
P(Make)                            0.4902
P(Miss)                            0.5098
P(Three-pointer)                   0.5481
P(Two-pointer)                     0.4519


## Section 2 — Formula

**Binomial distribution** — probability of exactly $k$ successes in $n$ independent trials:

$$P(X = k) = \binom{n}{k} \, p^k \, (1-p)^{n-k}$$

| Symbol | Meaning |
|--------|----------------------------------|
| $n$    | number of trials (shots)         |
| $k$    | desired number of successes      |
| $p$    | probability of success per trial |

**Assumptions:** shots are independent and $p$ is constant across shots.

---

**Applied numerically:**

**Q1** — P(exactly 3 makes in 4 shots), with $p = 0.4902$:

$$P(X=3) = \binom{4}{3} \times 0.4902^3 \times 0.5098^1 = 4 \times 0.1177 \times 0.5098 \approx 0.2402$$

**Q2** — P(exactly 4 three-pointers in 5 shots), with $p = 0.5481$:

$$P(X=4) = \binom{5}{4} \times 0.5481^4 \times 0.4519^1 = 5 \times 0.0902 \times 0.4519 \approx 0.2039$$

## Section 2 — Probabilities
___

### Questions to Answer

1. What is the probability of Steph **making 3 of the next 4 shots**?
2. What is the probability that **4 of the next 5 shots are three-pointers**?

---



In [10]:
from scipy.stats import binom

# 1.What is the probability of making 3 of the next 4 shots?
# p_q1 = P(3 makes in 4 shots)
p = p_make                     # probability of success (make)
n_shots = 4                    # number of trials
n = n_shots
k_successes = 3                # number of successes
k = k_successes
p_q1 = binom.pmf(k, n, p)      # probability of exactly k successes in n trials
print(f"Probability of making exactly 3 out of the next 4 shots: {p_q1:.4f}")


Probability of making exactly 3 out of the next 4 shots: 0.2402


In [11]:
# 2. What is the probability that 4 of the next 5 shots are three-pointers?
# p_q2 = P(4 three-pointers in 5 shots)
p = p_three                    # probability of success (three-pointer)
n_shots = 5                    # number of trials
n = n_shots
k_successes = 4                # number of successes
k = k_successes
p_q2 = binom.pmf(k, n, p)      # probability of exactly k successes in n trials
print(f"Probability of making exactly 4 out of the next 5 shots as three-pointers: {p_q2:.4f}")

Probability of making exactly 4 out of the next 5 shots as three-pointers: 0.2039


In [12]:
print("=" * 55)
print(f"{'SECTION 2 — PROBABILITIES':^55}")
print("=" * 55)
print(f"{'Question':<40} {'Probability':>10}")
print("-" * 55)
print(f"{'P(exactly 3 makes in 4 shots)':<40} {p_q1:>10.4f}")
print(f"{'P(exactly 4 three-pointers in 5 shots)':<40} {p_q2:>10.4f}")
print("=" * 55)
print()
print("Assumptions:")
print("  - Shots are independent events")
print("  - p(make) and p(three) are constant across shots")
print(f"  - p(make)  = {p_make:.4f} (estimated from data)")
print(f"  - p(three) = {p_three:.4f} (estimated from data)")
print("=" * 55)

               SECTION 2 — PROBABILITIES               
Question                                 Probability
-------------------------------------------------------
P(exactly 3 makes in 4 shots)                0.2402
P(exactly 4 three-pointers in 5 shots)       0.2039

Assumptions:
  - Shots are independent events
  - p(make) and p(three) are constant across shots
  - p(make)  = 0.4902 (estimated from data)
  - p(three) = 0.5481 (estimated from data)


## Section 3 — Formula

**Conditional probability** — probability of $A$ given that $B$ already occurred:

$$P(A \mid B) = \frac{P(A \cap B)}{P(B)} = \frac{\text{count}(A \text{ and } B)}{\text{count}(B)}$$

The joint probability is calculated as:

$$P(A \cap B) = \frac{\text{count}(A \text{ and } B)}{N}$$

**Key rule:** the notation $P(A \mid B)$ means:
- **Left of** $\mid$ → what you want to discover
- **Right of** $\mid$ → evidence you already have

Applied in this section:

$$P(\text{make} \mid \text{three}) = \frac{P(\text{make} \cap \text{three})}{P(\text{three})}$$

$$P(\text{lead} \mid \text{three}) = \frac{P(\text{lead} \cap \text{three})}{P(\text{three})}$$

$$P(\text{make} \mid \text{two}) = \frac{P(\text{make} \cap \text{two})}{P(\text{two})}$$

$$P(\text{lead} \mid \text{two}) = \frac{P(\text{lead} \cap \text{two})}{P(\text{two})}$$

## Section 3 — Conditional Probabilities (Future)

### Questions to Answer

If the next shot Steph shoots is a **three-pointer**:
1. What is the probability he **makes** it?
2. What is the probability it was taken while his team had the **lead**?

If the next shot Steph shoots is a **two-pointer**:

3. What is the probability he **makes** it?
4. What is the probability it was taken while his team had the **lead**?

---

**Ex1: What is the probability of making a shot given that it is a three-pointer?**

In [13]:
# Section 3: Conditional Probabilities
# Ex1: What is the probability of making a shot given that it is a three-pointer?
# P(make | three) = P(make ∩ three) / P(three)
# Step 1: P(make ∩ three) — joint probability
count_make_and_three = df[(df['result']) & (df['shot_type'] == 3)].shape[0]
p_make_and_three = count_make_and_three / total_shots

In [14]:
# Step 2: Apply conditional probability formula
p_make_given_three = p_make_and_three / p_three

print(f"Count(make AND three): {count_make_and_three}")
print(f"P(make ∩ three):        {p_make_and_three:.4f}")
print(f"P(three):               {p_three:.4f}")
print(f"P(make | three):        {p_make_given_three:.4f}")

Count(make AND three): 329
P(make ∩ three):        0.2294
P(three):               0.5481
P(make | three):        0.4186


In [15]:
# Section 3: Conditional Probabilities
# Ex2: What is the probability of lead given the shot is three-pointer ?
# P(lead | three) = P(lead ∩ three) / P(three)
# Step 1: P(lead ∩ three ) - joint probability
count_lead_and_three = df[df['lead'] & (df['shot_type'] == 3)].shape[0]
p_lead_given_three = count_lead_and_three / count_shot_type[3]

print(f"P(lead | three): {p_lead_given_three:.4f}")


P(lead | three): 0.4936


In [16]:
# Step 2: Apply conditional probability formula
p_lead_and_three = count_lead_and_three / total_shots

print(f"Count(lead AND three): {count_lead_and_three}")
print(f"P(lead ∩ three):        {p_lead_and_three:.4f}")
print(f"P(three):               {p_three:.4f}")
print(f"P(lead | three):        {p_lead_given_three:.4f}")

Count(lead AND three): 388
P(lead ∩ three):        0.2706
P(three):               0.5481
P(lead | three):        0.4936


## Section 4 — Formula

**Bayes' Theorem** — inverts the conditional: you have $P(B \mid A)$ but want $P(A \mid B)$:

$$P(A \mid B) = \frac{P(B \mid A) \cdot P(A)}{P(B)}$$

| Term | Name | Meaning |
|------|------|---------|
| $P(A)$ | **Prior** | Belief about $A$ before seeing evidence |
| $P(B \mid A)$ | **Likelihood** | Probability of evidence given cause |
| $P(B)$ | **Evidence** | Total probability of observing $B$ |
| $P(A \mid B)$ | **Posterior** | Updated belief after evidence |

Applied in this section:

$$P(\text{three} \mid \text{make}) = \frac{P(\text{make} \mid \text{three}) \cdot P(\text{three})}{P(\text{make})}$$

$$P(\text{two} \mid \text{make}) = \frac{P(\text{make} \mid \text{two}) \cdot P(\text{two})}{P(\text{make})}$$

> **Why Bayes?** In Section 3 you had the evidence (shot type) and wanted the outcome.
> Here the direction is **reversed**: you know the outcome (make) and want to infer the shot type.

## Section 4 — Conditional Probabilities (Past)

### Questions to Answer

If Steph **just made a shot**:
1. What is the probability that it was a **three-pointer**?
2. What is the probability that it was a **two-pointer**?

> **Note:** The project says you *could* answer these by direct counting,
> but you **must** use Bayes' Theorem and show your steps.

---

In [17]:
# Section 4: Conditional Probabilities (Past) — Bayes' Theorem
# P(three | make) = P(make | three) * P(three) / P(make)
# P(two   | make) = P(make | two)   * P(two)   / P(make)

# Step 1: P(make | two) — likelihood needed for Ex2 (not yet calculated)
count_make_and_two = df[df['result'] & (df['shot_type'] == 2)].shape[0]
p_make_and_two = count_make_and_two / total_shots
p_make_given_two = p_make_and_two / p_two

print(f"Count(make AND two):  {count_make_and_two}")
print(f"P(make | two):        {p_make_given_two:.4f}")
print(f"P(make | three):      {p_make_given_three:.4f}  (from Section 3)")

Count(make AND two):  374
P(make | two):        0.5772
P(make | three):      0.4186  (from Section 3)


In [18]:
# Step 2: Apply Bayes' Theorem
# Ex1: P(three | make) = P(make | three) * P(three) / P(make)
p_three_given_make = (p_make_given_three * p_three) / p_make

# Ex2: P(two | make) = P(make | two) * P(two) / P(make)
p_two_given_make = (p_make_given_two * p_two) / p_make

print(f"P(three | make): {p_three_given_make:.4f}")
print(f"P(two   | make): {p_two_given_make:.4f}")

P(three | make): 0.4680
P(two   | make): 0.5320


In [19]:
print("=" * 60)
print(f"{'SECTION 4 — CONDITIONAL PROBABILITIES (PAST)':^60}")
print("=" * 60)
print(f"\n{'Bayes Theorem components':}")
print(f"  P(make | three) = {p_make_given_three:.4f}  (likelihood)")
print(f"  P(make | two)   = {p_make_given_two:.4f}  (likelihood)")
print(f"  P(three)        = {p_three:.4f}  (prior)")
print(f"  P(two)          = {p_two:.4f}  (prior)")
print(f"  P(make)         = {p_make:.4f}  (evidence)")
print()
print(f"{'Question':<40} {'Probability':>10}")
print("-" * 55)
print(f"{'P(three | make)':<40} {p_three_given_make:>10.4f}")
print(f"{'P(two   | make)':<40} {p_two_given_make:>10.4f}")
print("=" * 60)
print(f"\nSanity check — P(three|make) + P(two|make) = {p_three_given_make + p_two_given_make:.4f} (must be 1.0)")

        SECTION 4 — CONDITIONAL PROBABILITIES (PAST)        

Bayes Theorem components
  P(make | three) = 0.4186  (likelihood)
  P(make | two)   = 0.5772  (likelihood)
  P(three)        = 0.5481  (prior)
  P(two)          = 0.4519  (prior)
  P(make)         = 0.4902  (evidence)

Question                                 Probability
-------------------------------------------------------
P(three | make)                              0.4680
P(two   | make)                              0.5320

Sanity check — P(three|make) + P(two|make) = 1.0000 (must be 1.0)


### Ex3 — Retrospective: P(lead | make)

> **Retrospective probability:** we already know the outcome (Steph *made* the shot) 
and are inferring the context (was the team leading?) — reasoning backward from result to circumstance.

---

**Terms needed:**

| Term | Name | Value |
|------|------|-------|
| $P(\text{lead})$ | Prior | proportion of shots taken while leading |
| $P(\text{make} \mid \text{lead})$ | Likelihood | make rate when team was leading |
| $P(\text{make})$ | Evidence | overall make rate |

**Via Bayes' Theorem:**

$$P(\text{lead} \mid \text{make}) = \frac{P(\text{make} \mid \text{lead}) \cdot P(\text{lead})}{P(\text{make})}$$

**Equivalently, via direct joint probability** (both give the same result):

$$P(\text{lead} \mid \text{make}) = \frac{P(\text{make} \cap \text{lead})}{P(\text{make})} = \frac{\text{count}(\text{make AND lead})}{\text{count}(\text{make})}$$

In [20]:
# Ex3: P(lead | make) — retrospective conditional probability
# Approach A — Bayes: P(lead|make) = P(make|lead) * P(lead) / P(make)
# Approach B — Direct joint: P(lead|make) = P(make ∩ lead) / P(make)

# Step 1: compute components
count_lead          = df['lead'].sum()
count_make_and_lead = df[df['result'] & df['lead']].shape[0]

p_lead              = count_lead / total_shots                    # prior
p_make_given_lead   = count_make_and_lead / count_lead            # likelihood
p_make_and_lead     = count_make_and_lead / total_shots           # joint

# Step 2: apply both approaches
p_lead_given_make_bayes  = (p_make_given_lead * p_lead) / p_make  # Bayes
p_lead_given_make_direct = p_make_and_lead / p_make               # direct joint

print("Bayes components:")
print(f"  P(lead):           {p_lead:.4f}  (prior)")
print(f"  P(make | lead):    {p_make_given_lead:.4f}  (likelihood)")
print(f"  P(make):           {p_make:.4f}  (evidence)")
print(f"  P(make ∩ lead):    {p_make_and_lead:.4f}  (joint)")
print()
print(f"P(lead | make) via Bayes:       {p_lead_given_make_bayes:.4f}")
print(f"P(lead | make) via direct joint:{p_lead_given_make_direct:.4f}")
print(f"Sanity check — both match: {abs(p_lead_given_make_bayes - p_lead_given_make_direct) < 1e-10}")

Bayes components:
  P(lead):           0.5035  (prior)
  P(make | lead):    0.5471  (likelihood)
  P(make):           0.4902  (evidence)
  P(make ∩ lead):    0.2755  (joint)

P(lead | make) via Bayes:       0.5619
P(lead | make) via direct joint:0.5619
Sanity check — both match: True


In [21]:
# P(make | two) = P(make ∩ two) / P(two)
count_make_and_two = df[df['result'] & (df['shot_type'] == 2)].shape[0]
p_make_and_two = count_make_and_two / total_shots
p_make_given_two = p_make_and_two / p_two

print(f"Count(make AND two): {count_make_and_two}")
print(f"P(make ∩ two):       {p_make_and_two:.4f}")
print(f"P(two):              {p_two:.4f}")
print(f"P(make | two):       {p_make_given_two:.4f}")

Count(make AND two): 374
P(make ∩ two):       0.2608
P(two):              0.4519
P(make | two):       0.5772


In [22]:
# P(lead | two) = P(lead ∩ two) / P(two)
count_lead_and_two = df[df['lead'] & (df['shot_type'] == 2)].shape[0]
p_lead_and_two = count_lead_and_two / total_shots
p_lead_given_two = p_lead_and_two / p_two

print(f"Count(lead AND two): {count_lead_and_two}")
print(f"P(lead ∩ two):       {p_lead_and_two:.4f}")
print(f"P(two):              {p_two:.4f}")
print(f"P(lead | two):       {p_lead_given_two:.4f}")

Count(lead AND two): 334
P(lead ∩ two):       0.2329
P(two):              0.4519
P(lead | two):       0.5154


## Section 4 — Conclusions

- **P(Three | Make) = 0.4680** — when Steph makes a shot, there is a ~47% chance it was a three-pointer.
- **P(Two | Make) = 0.5320** — when Steph makes a shot, there is a ~53% chance it was a two-pointer.
- **P(Lead | Make) = 0.5619** — when Steph makes a shot, there is a ~56% chance his team was leading at the time.

These values match exactly the conditional probabilities calculated as intermediate terms in Section 3, which is a strong internal consistency check confirming that all results are coherent across sections.

Even though Steph attempts more three-pointers — P(Three) = 0.5481 — made shots are slightly more likely to be two-pointers. The reason is that his two-point accuracy is meaningfully higher: P(Make | Two) = 0.5772 versus P(Make | Three) = 0.4186. The higher conversion rate on two-pointers offsets the lower attempt volume, shifting the posterior toward two-pointers given a make.